lib/OpenQA/Worker/Jobs.pm

```diff
@@ -164,6 +175,21 @@ sub upload {
 
     my $res = $OpenQA::Worker::Common::ua->start($tx);
 
+    # Upload known server failures (Instead of anything that's not 200)
+    unless ($res->res->code =~ m/^5/) {
+        if ($retry_counter++ == $retry_limit) {
+            my $msg = sprintf("Uploads have failed for %s\n", $filename);
+            open(my $log, '>>', "autoinst-log.txt");
+            print $log $msg;
+            close $log;
+            print STDERR $msg;
+            return 0;
+        }
+        else {
+            goto START_UPLOAD;
```

Comment:
The `goto` on the other hand is very very non-idiomatic Perl, always use a loop construct instead.
```
while ($retry_counter < $retry_limit) {
  ...
  next;
  ...
}
```